<a href="https://colab.research.google.com/github/YunjiSon/2022spring/blob/main/Tensorflow_Hub.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Image classification

fine-tuning: AI에 대해 입력벡터, 출력벡터가 있고 중간에 행렬이 있는 것.
pretrained된 모델이 있다고 쳤을 때, fine tuning은 그 모델을 좀 확장시켰다. (나의 데이터를 가지고 조금 더 변경시키고, 확장시킨다.)

In [1]:
import tensorflow.compat.v2 as tf
import tensorflow_hub as hub

In [2]:
model = hub.KerasLayer("https://tfhub.dev/google/imagenet/inception_v3/classification/5")
#inception3에서 만드는것. 모델을 keraslayer라는 걸로 이렇게 가져오는 것.

In [10]:
from PIL import Image
import requests
url = 'http://images.cocodataset.org/val2017/000000039769.jpg'
#이미지 가져오기
image = Image.open(requests.get(url, stream=True).raw)

In [11]:
import numpy as np
image = np.array(image)
# reshape into shape [batch_size, height, width, num_channels]
image = tf.reshape(image, [1, image.shape[0], image.shape[1], image.shape[2]])
#3차원의 array를 reshape로 4차원으로 바꿔주기 이 모델이 4차원을 받아주는 모델이라.
# Use `convert_image_dtype` to convert to floats in the [0,1] range.
#어두우면 0, 밝으면 1 뭐 이런식으로.
image = tf.image.convert_image_dtype(image, tf.float32)

In [8]:
image[:,:,0]

<tf.Tensor: shape=(1, 480, 3), dtype=float32, numpy=
array([[[0.54901963, 0.09803922, 0.21960786],
        [0.5411765 , 0.08627451, 0.22352943],
        [0.5294118 , 0.08627451, 0.16470589],
        ...,
        [0.9294118 , 0.3921569 , 0.74509805],
        [0.90196085, 0.32941177, 0.86666673],
        [0.9333334 , 0.3921569 , 0.6862745 ]]], dtype=float32)>

In [12]:
logits = model(image)
#여기서 image가 벡터가 되는 것.
np.argmax(logits) #몇번째가 제일 높은 값이냐?

282

-->inception v3에 있는 데이터들이 1000개 있는데 그거 중에 282번째거가 가장 높다.
0번부터니까 엑셀에 적용하면 283번째 찾으면 된다.